op/op.go

```diff
@@ -52,12 +52,39 @@ func New(id string, result map[string]any, status Status, err error, params map[
 
 // StatusCounts takes a slice of op references and returns a map containing sums of each Status
 func StatusCounts(ops map[string]Op) (map[Status]int, error) {
-	statuses := make(map[Status]int)
-	for _, o := range ops {
-		if o.Status == "" {
-			return nil, fmt.Errorf("unable to build Statuses map, op not run: op=%s", o.Identifier)
-		}
-		statuses[o.Status]++
+	// copy our input into a new map that conforms to our Walk input type
+	m := make(map[string]any, len(ops))
+	for k, v := range ops {
+		m[k] = v
+	}
+
+	// Create our accumulator and Walk ops
+	acc := make(map[Status]int, 0)
+	statuses := WalkStatuses(acc, m)
+
+	if val, ok := statuses[""]; ok {
+		return nil, fmt.Errorf("op.StatusCounts received ops that did not have a status, count=%d", val)
 	}
 	return statuses, nil
 }
+
+// WalkStatuses performs a depth-first search of a tree of results and accumulates a map of status counts
+func WalkStatuses(statuses map[Status]int, results map[string]any) map[Status]int {
```

Comment:
We're mutating `statuses` here, and then returning it at the end. Perhaps we could avoid the mutation, though. What if `WalkStatuses` took only a `results` map, and it had an inner function defined, which would take a status map. We could recurse on the inner function and then return what it built up at the end of `WalkStatuses`. This would, I think, both make the calling of the `WalkStatuses` easier (no need to create an accumulator) and also avoid accidental mutation.

Reply:
`walkStatuses` could also just be an unexported function as well - the main idea is, can we avoid mutation in the public API?

Reply:
Not certain it's the ideal implementation, but it does let us pass in partially constructed accumulators for testing or for more elaborate concurrent reducers. You could reasonably do the same thing without passing in an accumulator though, treating it as the steps: `split/partition results` -> `map over result subset and return map[Status]int` -> `merge all map[Status]int`. Also, The accumulator is passede by so we don't have to worry about changes from outside goroutines.

Reply:
I was mistaken! maps are like channels not slices and get passed by reference even without an explicit pointer!

Reply:
Thanks -- used @nwchandler 's inner-closure version of this function. Let me know if that's good to go!